protocol/consistencychecks.go

```diff
@@ -42,222 +38,243 @@ type ConsistencyChecks struct {
-	if err == PassedWithAProofOfAbsence ||
-		err == PassedWithAProofOfInclusion {
-		cc.SavedSTR = str
+func (cc *ConsistencyChecks) checkConsistency(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	var err error
+	switch requestType {
+	case RegistrationType:
+		err = cc.verifyRegistration(requestType, msg, uname, key)
+	case KeyLookupType:
+		err = cc.verifyKeyLookup(requestType, msg, uname, key)
+	default:
+		panic("[coniks] Unknown request type.")
 	}
-
 	return err.(ErrorCode)
 }
 
 func (cc *ConsistencyChecks) verifyRegistration(requestType int,
-	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
-
-	df, ok := msg.DirectoryResponse.(*DirectoryProof)
-	if !ok {
-		return nil, ErrorMalformedDirectoryMessage
-	}
-
+	msg *Response, uname string, key []byte) error {
+	df := msg.DirectoryResponse.(*DirectoryProof)
 	ap := df.AP
 	str := df.STR
+	tb := df.TB
 
-	// 1. verify STR
-	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
-		return nil, err
-	}
-
-	// 2. verify Auth path
-	proofType, err := verifyAuthPath(uname, key, ap, str)
-	if err != nil {
-		return nil, err
-	}
-
-	// 3. verify (response code, proof type) pair
+	proofType := ap.ProofType()
 	switch {
-	case msg.Error == ErrorNameExisted:
-	case msg.Error == Success && proofType == proofOfAbsence:
+	case msg.Error == ErrorNameExisted && proofType == m.ProofOfInclusion && tb == nil:
+	case msg.Error == ErrorNameExisted && proofType == m.ProofOfAbsence:
+	case msg.Error == Success && proofType == m.ProofOfAbsence:
 	default:
-		return nil, ErrorMalformedDirectoryMessage
+		return ErrorMalformedDirectoryMessage
 	}
 
-	// 4. verify returned promise
-	if err := cc.verifyReturnedPromise(requestType,
-		msg.Error, df, uname, key); err != nil {
-		return nil, err
+	if err := verifyAuthPath(uname, key, ap, str); err != nil {
+		return err
 	}
 
-	if proofType == proofOfAbsence {
-		return str, PassedWithAProofOfAbsence
-	}
-	return str, PassedWithAProofOfInclusion
+	return Passed
 }
 
 func (cc *ConsistencyChecks) verifyKeyLookup(requestType int,
-	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
-
-	df, ok := msg.DirectoryResponse.(*DirectoryProof)
-	if !ok {
-		return nil, ErrorMalformedDirectoryMessage
-	}
-
+	msg *Response, uname string, key []byte) error {
+	df := msg.DirectoryResponse.(*DirectoryProof)
 	ap := df.AP
 	str := df.STR
+	tb := df.TB
 
-	// 1. verify STR
-	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
-		return nil, err
-	}
-
-	// 2. verify Auth path
-	proofType, err := verifyAuthPath(uname, key, ap, str)
-	if err != nil {
-		return nil, err
-	}
-
-	// 3. verify (response code, proof type) pair
+	proofType := ap.ProofType()
 	switch {
-	case msg.Error == ErrorNameNotFound && proofType == proofOfAbsence:
-	case msg.Error == Success:
+	case msg.Error == ErrorNameNotFound && proofType == m.ProofOfAbsence:
+	case msg.Error == Success && proofType == m.ProofOfInclusion && tb == nil:
```

Comment:
This case is going to change when we support key changes, so I would add a FIXME or TODO to indicate this.